core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala

```diff
@@ -86,10 +89,115 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
+   * else if attribute name and value is specified, subset match is performed on slave attributes
+   */
+  private[mesos] def matchesAttributeRequirements(
+    slaveOfferConstraints: Map[String, Set[String]],
+    offerAttributes: Map[String, Set[String]]): Boolean =
+    if (slaveOfferConstraints.isEmpty) {
+      true
+    } else {
+      slaveOfferConstraints.forall {
+        // offer has the required attribute and subsumes the required values for that attribute
+        case (name, requiredValues) =>
+          // The attributes and their values are case sensitive during comparison
+          // i.e tachyon -> true != Tachyon -> true != tachyon -> True
+          offerAttributes.contains(name) && requiredValues.subsetOf(offerAttributes(name))
+
+      }
+    }
+
+  /**
+   * Parses the attributes constraints provided to spark and build a matching data struct:
+   *  Map[<attribute-name>, Set[values-to-match]
+   *  The constraints are specified as ';' separated key-value pairs where keys and values
+   *  are separated by ':'. The ':' implies equality. For example:
+   *  {{{
+   *  parseConstraintString("tachyon:true;zone:us-east-1a,us-east-1b")
+   *  // would result in
+   *  <code>
+   *  Map(
+   *    "tachyon" -> Set("true"),
+   *    "zone":   -> Set("us-east-1a", "us-east-1b")
+   *  )
+   *  }}}
+   * @param constraintsVal constaints string consisting of ';' separated key-value pairs (separated
+   *                       by ':')
+   * @return  Map of constraints to match resources offers.
+   */
+  private[mesos] def parseConstraintString(constraintsVal: String): Map[String, Set[String]] = {
+    /*
+      Based on mesos docs:
+      attributes : attribute ( ";" attribute )*
+      attribute : labelString ":" ( labelString | "," )+
+      labelString : [a-zA-Z0-9_/.-]
+    */
+    val splitter = Splitter.on(';').trimResults().withKeyValueSeparator(':')
+    // kv splitter
+    if (constraintsVal.isEmpty) {
+      Map()
+    } else {
+      try {
+        Map() ++ mapAsScalaMap(splitter.split(constraintsVal)).map {
+          case (k, v) =>
+            if (v == null) {
+              (k, Set[String]())
+            } else {
+              (k, v.split(',').toSet)
+            }
+        }
+      } catch {
+        case e: Throwable =>
+          throw new IllegalArgumentException(s"Bad constraint string: $constraintsVal", e)
+      }
+    }
+  }
+
+  /**
+   * For the list of offers received, find the ones that match the offer constraints (if specified)
+   * @param offers set of all offers received
+   * @return Offers that match the constraints
+   */
+  private[mesos] def filterOffersByConstraints(
+    offers: JList[Offer],
+    offerConstraints: Map[String, Set[String]]): mutable.Buffer[Offer] = offers.filter { o =>
+    matchesAttributeRequirements(offerConstraints, (o.getAttributesList map getAttribute).toMap)
+  }
+
+  // These defaults copied from YARN
+  val OVERHEAD_FRACTION = 0.10
```

Comment:
If you want to move the memory stuff here, I think you should change the vals to prefix MEMORY_ so we know what OVERHEADs these are for.